src/generator.rs

```diff
@@ -220,8 +234,119 @@ impl<'a> Generator<'a> {
                 ch.name
             ))));
         };
+
+        let mut body = String::new();
+
+        if self.config.epub_version == Some(3) && self.config.footnote_backrefs {
+            body.push_str(&self.render_with_footnote_backrefs(chapter_dir, ch));
```

Comment:
Looks like we are repeating ourselves inside - `render_with_footnote_backrefs(.....)` making 
either flow for `epub_version == Some(3) && self.config.footnote_backrefs`
or alternative flow doing mostly the same functionality.

It would be good to have an 'optional step' in case conditions (epub_version + footnote_backrefs') for updating page content with all those functional (extracted into another component probably). Otherwise code becomes too long, compilcate and 'durty'... agree with @dieterplex 

Looks like `fn footnote_reference_event<'b>` can be extracted into separate module at least, something like 'footnote.rs'.

If you have complexities to refactor, I will do that after a merging PR into master. We need to do that before release new version.

Reply:
It would be better if you or someone tackled the refactoring... I agree about the repeated function being bad style, but I couldn't resolve the compiler errors when I tried to refactor it. Rust is cool 😎 but sometimes it becomes a 4D-chess puzzle.